tests/src/tests.rs

```diff
@@ -585,67 +661,172 @@ fn test_part(
         frames.clear();
     }
 
-    // Map diagnostics to range and message format, discard traces and errors from
-    // other files, collect hints.
-    //
-    // This has one caveat: due to the format of the expected hints, we can not
-    // verify if a hint belongs to a diagnostic or not. That should be irrelevant
-    // however, as the line of the hint is still verified.
-    let mut actual_diagnostics = HashSet::new();
-    for diagnostic in &diagnostics {
-        // Ignore diagnostics from other files.
-        if diagnostic.span.id().map_or(false, |id| id != source.id()) {
-            continue;
-        }
+    if !metadata.invalid_data.is_empty() {
+        ok = false;
+        writeln!(output, "  Subtest {i} has invalid metadata, failing the test:")
+            .unwrap();
+        for (annotation, error) in &metadata.invalid_data {
+            if let Some(annotation) = annotation {
+                write!(output, "    {error}|").unwrap();
 
-        let annotation = Annotation {
-            kind: match diagnostic.severity {
-                Severity::Error => AnnotationKind::Error,
-                Severity::Warning => AnnotationKind::Warning,
-            },
-            range: world.range(diagnostic.span),
-            message: diagnostic.message.replace("\\", "/"),
-        };
-
-        if validate_hints {
-            for hint in &diagnostic.hints {
-                actual_diagnostics.insert(Annotation {
-                    kind: AnnotationKind::Hint,
-                    message: hint.clone(),
-                    range: annotation.range.clone(),
-                });
+                // line is a logic error here, what should I do with it?
+                print_annotation(output, &source, line, annotation)
+            } else {
+                writeln!(output, "    {error}").unwrap();
             }
         }
-
-        actual_diagnostics.insert(annotation);
     }
 
-    // Basically symmetric_difference, but we need to know where an item is coming from.
-    let mut unexpected_outputs = actual_diagnostics
-        .difference(&metadata.annotations)
-        .collect::<Vec<_>>();
-    let mut missing_outputs = metadata
+    // we never check autocomplete and error at the same time
+
+    let diagnostic_annotations = metadata
         .annotations
-        .difference(&actual_diagnostics)
-        .collect::<Vec<_>>();
+        .iter()
+        .filter(|a| {
+            !matches!(
+                a.kind,
+                AnnotationKind::AutocompleteContains
+                    | AnnotationKind::AutocompleteExcludes
+            )
+        })
+        .cloned()
+        .collect::<HashSet<_>>();
+
+    if !validate_autocomplete {
```

Comment:
This function is _very_ long now and it also contains a fair bit of duplication. I think it needs some subfunctions and maybe that also helps with the duplication.

Reply:
I split it and it's much more readable now, I got rid of some duplication as well with a new `InvalidMetadata::write` helper function